celery/app/trace.py

```diff
@@ -209,6 +210,16 @@ def handle_failure(self, task, req, store_errors=True, call_errbacks=True):
                                       traceback=tb,
                                       einfo=einfo)
             self._log_error(task, req, einfo)
+            return einfo
+        except Exception as fail_exc:
+            _, _, tmp_tb = sys.exc_info()
+            try:
+                msg = 'Exception raised while handling failure: {0!r}:\n{1}' \
+                    .format(fail_exc, tmp_tb)
+                warn(RuntimeWarning(msg))
```

Comment:
I suspect we want to use `logger.exception` instead of the Python warnings system. That way if we have serious problems storing failures the errors will at least reliably make it to Sentry. Otherwise I think we'll only log the warning once and then suppress it.

Reply:
There's a test in `t/unit/worker/test_request.py` that fails if I don't emit a `RuntimeWarning`:

```python
    def test_trace_catches_exception(self):

        @self.app.task(request=None, shared=False)
        def raising():
            raise KeyError('baz')

        with pytest.warns(RuntimeWarning):
            res = trace_task(raising, uuid(), [], {}, app=self.app)[0]
>           assert isinstance(res, ExceptionInfo)
E           Failed: DID NOT WARN. No warnings of type (<class 'RuntimeWarning'>,) was emitted. The list of emitted warnings is: [].
```

I basically took the code snippet from here and modified it: https://github.com/sightmachine/celery/blob/c5793740685fa1376b2d06f3678c872ca175ed6f/celery/app/trace.py#L535

Reply:
That's fine, we can report the warning if we're required to. I think we should still log the error via `self._log_error` though. Otherwise Python will only print the warning the first time it occurs and the error will not be surfaced in Sentry — we'd effectively have no way to debug problems with the backend.

On a side note, why duplicate the code instead of calling `report_internal_error` directly? It seems like we have all of the information that we need in `handle_failure` to call the existing method instead of making the tests rely on having the same code copy-pasted in two places.

Reply:
Actually, it turns out a single call to `self._log_error` takes care of the test passing so that's what I'll do.

```python
        except Exception as fail_exc:
            _, _, tmp_tb = sys.exc_info()
            try:
                self._log_error(task, req, fail_exc)
            finally:
                del tmp_tb
```

Thanks for the suggestion, looks much better now!

Reply:
Wait, how did the test pass if we're calling `self._log_error`? Shouldn't the test be failing since we're not generating the warning?

Reply:
@amastilovic: Looking at the code it appears that the call to `self._log_error` is incorrect. `_log_error` expects an `ExceptionInfo` object but an exception is being passed. That's causing the code to throw an exception — so the merged changes aren't actually making `handle_failure` catch exceptions.

Reply:
@bremac you are absolutely correct, the code expects an `ExceptionInfo` and will fail as it is. At first I thought my code changes weren't really taken by `pytest` since I had `celery-4.1.0+sm3` installed while I was preparing the package for JFrog.

Now that I've got all that under control I can confidently report that `test_Request.test_trace_catches_exception` passes with `self._log_error(task, req, fail_exc)` and does not report any issue :-) In fact, if I replace the line `self._log_error(task, req, fail_exc)` with `print("foo")` the test fails.

So that test is closest to what Celery has to cover this issue, and it doesn't really seem to care about `which` errors were generated as long as they were generated.

```python
    def test_trace_catches_exception(self):

        @self.app.task(request=None, shared=False)
        def raising():
            raise KeyError('baz')

        with pytest.warns(RuntimeWarning):
            res = trace_task(raising, uuid(), [], {}, app=self.app)[0]
            assert isinstance(res, ExceptionInfo)
```

Those are some serious bug spotting skills!!!

Anyway I've created this PR to fix the issue: https://github.com/sightmachine/celery/pull/7